### PR TITLE
Deduplicate hashing code

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
Inspired by some ansel code introducing dt_hash(), modified code with const and pointer type casting.

Use new inline function all over the code for code deduplication and readability. In two places the hash calculation was not of best quality.

### CODE REVIEW

File: src/common/darktable.h [hunk 1]
```diff
@@ -652,6 +652,20 @@ static inline int dt_get_thread_num()
 #endif
 }
 
+#define DT_INITHASH 5381
+// idea from ansel code @aurelienpierre
```
Comment (TurboGit) [inline src/common/darktable.h:656]: Well we are talking about a simple code refactoring. No need to this comment.

**Comment by TurboGit** [comment]
BTW, maybe this can also be used in `dt_view_get_view_context`? We could cast all variables as uint32_t and create an array with those values then call `dt_hash()`. What do you think?

**Comment by jenshannoschwalm** [comment]
Yes I spotted that too. Will have closer look

**Comment by jenshannoschwalm** [comment]
> create an array with those values then call `dt_hash()`

Right. We could also use the parameters directly to calc the hash but this way it's safe.

File: src/views/view.c [hunk 2]
```diff
@@ -1804,20 +1802,17 @@ dt_view_context_t dt_view_get_view_context(void)
...
-  ADD_TO_CONTEXT((int)(zoom_x * flt_prec));
-  ADD_TO_CONTEXT((int)(zoom_y * flt_prec));
-  ADD_TO_CONTEXT(iso_12646);
-  ADD_TO_CONTEXT(focus_peaking);
-
-  // dt_print(DT_DEBUG_EXPOSE, "dt_view_get_view_context ctx==%" PRIx64 " scale=%f\n", ctx, zoom_scale);
-  return ctx;
+  // calculate a hash on view parameters. Use flt_prec here to avoid different hashes
+  // for irrelevant variations for the zooms.
+  const float flt_prec = 1.e6;
```
Comment (TurboGit) [inline src/views/view.c:1808]: looks perfect!

**Comment by TurboGit** [comment]
@jenshannoschwalm : I'll merge this first and will adapt #15910 to use the new `dt_view_get_view_context` implementation.

**Comment by jenshannoschwalm** [comment]
Right. While being here, wouldn't it be better to define & use `dt_hash_t` as `uint64_t` for consistence? Also `view_context` is just a hash, so naming might be better as `dt_view_get_hash` ?

File: src/common/darktable.h [hunk 3]
```diff
@@ -652,6 +652,19 @@ static inline int dt_get_thread_num()
 #endif
 }
 
+#define DT_INITHASH 5381
+static inline uint64_t dt_hash(uint64_t hash, const void *data, const size_t size)
```
Comment (TurboGit) [inline src/common/darktable.h:656]: create a new type d_hash_t and define dt_hash as:

```
typedef uint64_t dt_hash_t;

static inline dt_hash_t dt_hash(dt_hash_t hash, const void *data, const size_t size);
```

File: src/develop/develop.c [hunk 4]
```diff
@@ -3262,7 +3262,7 @@ uint64_t dt_dev_hash_plus(dt_develop_t *dev,
                           const double iop_order,
                           const dt_dev_transform_direction_t transf_direction)
 {
-  uint64_t hash = 5381;
+  uint64_t hash = DT_INITHASH;
```
Comment (TurboGit) [inline src/develop/develop.c:3265]: `uint64_t` -> `dt_hash_t`

File: src/develop/develop.c [hunk 5]
```diff
@@ -3374,7 +3374,7 @@ uint64_t dt_dev_hash_distort_plus(dt_develop_t *dev,
                                   const double iop_order,
                                   const dt_dev_transform_direction_t transf_direction)
 {
-  uint64_t hash = 5381;
+  uint64_t hash = DT_INITHASH;
```
Comment (TurboGit) [inline src/develop/develop.c:3377]: likewise and in all other places

**Review by TurboGit — COMMENTED** [review]
> Right. While being here, wouldn't it be better to define & use dt_hash_t as uint64_t for consistence?

Ah.... Just read that and I had prepared a review with the exact same idea :)

**Comment by TurboGit** [comment]
> Also `view_context` is just a hash, so naming might be better as `dt_view_get_hash` ?

Yes, agreed or `dt_view_get_context_hash` has this is really the idea, getting the whole **view context** to be able to detect changes and so trigger recomputing.

**Comment by jenshannoschwalm** [comment]
Will do tonight.

File: src/iop/lens.cc [hunk 6]
```diff
@@ -242,7 +242,7 @@ typedef struct dt_iop_lens_data_t
   float v_steepness;
   float reserved[2];
   float vigspline[VIGSPLINES];
-  uint32_t vighash;
+  dt_hash_t vighash;
```
Comment (jenshannoschwalm) [inline src/iop/lens.cc:245]: @TurboGit for the record, this is a minor bug but i will prepare a pr just for this and 4.6.1

**Review by TurboGit — APPROVED** [review]
Thanks!

**Pull request merged**